File: sdk/csharp/src/Providers/Inboxes.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text.Json.Nodes;

namespace XxxXTeam.TempMail.Providers;

/// <summary>Inboxes 渠道（inboxes.com/api/v2）。GET JSON API，逐封拉取详情。</summary>
public static class Inboxes
{
    private const string ApiBase = "https://inboxes.com/api/v2";
    private const string DefaultDomain = "blondmail.com";
    private static readonly Random Rand = new();

    private static readonly Dictionary<string, string> Headers = new()
    {
        ["Accept"] = "application/json",
        ["Origin"] = "https://inboxes.com",
        ["Referer"] = "https://inboxes.com/",
        ["User-Agent"] = "Mozilla/5.0",
    };

    private static string RandomLocal()
    {
        const string chars = "abcdefghijklmnopqrstuvwxyz0123456789";
        var sb = new System.Text.StringBuilder("sdk");
        for (var i = 0; i < 16; i++) sb.Append(chars[Rand.Next(chars.Length)]);
        return sb.ToString();
    }

    private static List<string> GetDomains()
    {
        var resp = Http.Get($"{ApiBase}/domain", Headers);
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body) as JsonObject;
        var items = data?["domains"] as JsonArray;
        var domains = new List<string>();
        if (items is not null)
            foreach (var item in items)
                if (item is JsonObject o)
                {
                    var qdn = (o["qdn"]?.GetValue<string>() ?? "").Trim();
                    if (qdn.Length > 0) domains.Add(qdn);
                }
        if (domains.Count == 0) throw new Exception("inboxes: no domains");
        return domains;
    }

    private static string PickDomain(List<string> domains, string? preferred)
    {
        var wanted = (preferred ?? "").Trim().TrimStart('@').ToLowerInvariant();
        if (wanted.Length > 0)
            foreach (var d in domains)
                if (d.ToLowerInvariant() == wanted) return d;
        if (domains.Contains(DefaultDomain)) return DefaultDomain;
        return domains[0];
    }

    private static Dictionary<string, object?> Flatten(JsonObject raw, string recipient)
    {
        return new Dictionary<string, object?>
        {
            ["id"] = raw["uid"]?.GetValue<string>() ?? "",
            ["from"] = Pick(raw, "sf", "f"),
            ["to"] = PickOr(raw, recipient, "ib"),
            ["subject"] = Pick(raw, "s"),
            ["text"] = Pick(raw, "text"),
            ["html"] = Pick(raw, "html"),
            ["preview_text"] = Pick(raw, "ph"),
            ["date"] = Pick(raw, "cr"),
            ["isRead"] = !string.IsNullOrEmpty(Pick(raw, "ru")),
            ["attachments"] = Json.ToRaw(raw["at"] as JsonArray),
        };
    }

    /// <summary>创建临时邮箱（可指定域名）</summary>
    public static EmailInfo Generate(string? domain = null)
    {
        var domains = GetDomains();
        var selected = PickDomain(domains, domain);
        return new EmailInfo("inboxes", $"{RandomLocal()}@{selected}");
    }

    private static JsonObject? FetchDetail(string uid)
    {
        try
        {
            var resp = Http.Get($"{ApiBase}/message/{WebUtility.UrlEncode(uid)}", Headers);
            if (!resp.Ok) return null;
            return Json.Parse(resp.Body) as JsonObject;
        }
        catch { return null; }
    }

    /// <summary>获取邮件列表，逐封补全详情</summary>
    public static List<Email> GetEmails(string email)
    {
        var address = (email ?? "").Trim();
        if (address.Length == 0) throw new Exception("inboxes: empty email");

        var resp = Http.Get($"{ApiBase}/inbox/{WebUtility.UrlEncode(address)}", Headers);
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body) as JsonObject;
        var result = new List<Email>();
        if (data?["msgs"] is not JsonArray rows) return result;

        foreach (var r in rows)
        {
            if (r is not JsonObject row) continue;
            var uid = (row["uid"]?.GetValue<string>() ?? "").Trim();
            var detail = uid.Length > 0 ? FetchDetail(uid) : null;
            result.Add(Normalize.NormalizeEmail(Flatten(detail ?? row, address), address));
        }
        return result;
    }

    private static string Pick(JsonObject o, params string[] keys)
    {
        foreach (var k in keys)
        {
            if (o[k] is JsonNode n)
            {
                var s = Json.NodeToString(n);
                if (!string.IsNullOrEmpty(s)) return s;
            }
        }
        return "";
    }

    private static string PickOr(JsonObject o, string fallback, params string[] keys)
    {
        var v = Pick(o, keys);
        return v.Length > 0 ? v : fallback;
    }
}

File: sdk/php/src/Providers/Inboxes.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * Inboxes 渠道实现 — https://inboxes.com
 *
 * 创建：GET /api/v2/domain 取域名（qdn），本地拼随机 local part；
 * 读信：GET /api/v2/inbox/{email} 取 msgs，逐封 GET /api/v2/message/{uid} 补全正文。
 */
final class Inboxes
{
    private const CHANNEL = 'inboxes';
    private const API_BASE = 'https://inboxes.com/api/v2';
    private const DEFAULT_DOMAIN = 'blondmail.com';

    /** @var array<string,string> */
    private const HEADERS = [
        'Accept' => 'application/json',
        'Origin' => 'https://inboxes.com',
        'Referer' => 'https://inboxes.com/',
        'User-Agent' => 'Mozilla/5.0',
    ];

    private static function randomLocal(): string
    {
        $chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
        $out = 'sdk';
        for ($i = 0; $i < 16; $i++) {
            $out .= $chars[random_int(0, strlen($chars) - 1)];
        }
        return $out;
    }

    /**
     * @return string[]
     */
    private static function getDomains(): array
    {
        $resp = HttpClient::get(self::API_BASE . '/domain', self::HEADERS);
        if ($resp->getStatusCode() >= 400) {
            throw new \RuntimeException('inboxes: 域名请求失败 ' . $resp->getStatusCode());
        }
        $data = HttpClient::json($resp);
        $items = is_array($data['domains'] ?? null) ? $data['domains'] : [];
        $domains = [];
        foreach ($items as $item) {
            if (is_array($item)) {
                $qdn = trim((string) ($item['qdn'] ?? ''));
                if ($qdn !== '') {
                    $domains[] = $qdn;
                }
            }
        }
        if ($domains === []) {
            throw new \RuntimeException('inboxes: no domains');
        }
        return $domains;
    }

    /**
     * @param string[] $domains
     */
    private static function pickDomain(array $domains, ?string $preferred): string
    {
        $wanted = strtolower(ltrim(trim((string) $preferred), '@'));
        if ($wanted !== '') {
            foreach ($domains as $domain) {
                if (strtolower($domain) === $wanted) {
                    return $domain;
                }
            }
        }
        if (in_array(self::DEFAULT_DOMAIN, $domains, true)) {
            return self::DEFAULT_DOMAIN;
        }
        return $domains[0];
    }

    /**
     * @param array<mixed> $raw
     * @return array<mixed>
     */
    private static function flatten(array $raw, string $recipient): array
    {
        return [
            'id' => $raw['uid'] ?? '',
            'from' => $raw['sf'] ?? ($raw['f'] ?? ''),
            'to' => $raw['ib'] ?? $recipient,
            'subject' => $raw['s'] ?? '',
            'text' => $raw['text'] ?? '',
            'html' => $raw['html'] ?? '',
            'preview_text' => $raw['ph'] ?? '',
            'date' => $raw['cr'] ?? '',
            'isRead' => (bool) ($raw['ru'] ?? false),
            'attachments' => is_array($raw['at'] ?? null) ? $raw['at'] : [],
        ];
    }

    public static function generate(?string $domain = null): EmailInfo
    {
        $domains = self::getDomains();
        $selected = self::pickDomain($domains, $domain);
        return new EmailInfo(self::CHANNEL, self::randomLocal() . '@' . $selected);
    }

    /**
     * @return array<mixed>|null
     */
    private static function fetchDetail(string $uid): ?array
    {
        try {
            $resp = HttpClient::get(self::API_BASE . '/message/' . rawurlencode($uid), self::HEADERS);
            if ($resp->getStatusCode() >= 400) {
                return null;
            }
            $data = HttpClient::json($resp);
            return $data !== [] ? $data : null;
        } catch (\Throwable) {
            return null;
        }
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email): array
    {
        $address = trim($email);
        if ($address === '') {
            throw new \InvalidArgumentException('inboxes: empty email');
        }
        $resp = HttpClient::get(self::API_BASE . '/inbox/' . rawurlencode($address), self::HEADERS);
        if ($resp->getStatusCode() >= 400) {
            throw new \RuntimeException('inboxes: 收件箱请求失败 ' . $resp->getStatusCode());
        }
        $data = HttpClient::json($resp);
        $rows = $data['msgs'] ?? null;
        if (!is_array($rows)) {
            return [];
        }
        $out = [];
        foreach ($rows as $row) {
            if (!is_array($row)) {
                continue;
            }
            $uid = trim((string) ($row['uid'] ?? ''));
            $detail = $uid !== '' ? self::fetchDetail($uid) : null;
            $out[] = Normalize::email(self::flatten($detail ?? $row, $address), $address);
        }
        return $out;
    }
}
